File: Java_Design-Patterns/src/facade/Point.java
package facade;

public class Point {
    private PointCarte pc; // 2. Design a "wrapper" class

    public Point(double x, double y) {
        this.pc = new PointCarte(x, y);
    }

    @Override
    public String toString() {
        return pc.toString();
    }

    // 4. Wrapper maps
    public void move(double dx, double dy) {
        pc.move(dx, dy);
    }

    public void rotate(int angle, Point point) {
        double x = pc.getX() - point.pc.getX();
        double y = pc.getY() - point.pc.getY();
        PointPolar pp = new PointPolar(Math.sqrt(Math.pow(x, 2) + Math.pow(y, 2)), Math.atan2(y, x) * 180 / Math.PI);

        // 4. Wrapper maps
        pp.rotate(angle);
        System.out.println("PointPolar is " + pp);
        String str = pp.toString();
        int i = str.indexOf("@");
        double r = Double.parseDouble(str.substring(1, i));
        double a = Double.parseDouble(str.substring(i + 1, str.length() - 1));

        pc = new PointCarte(r * Math.cos(a * Math.PI / 180) + point.pc.getX(),
                            r * Math.sin(a * Math.PI / 180) + point.pc.getY());
    }
}

File: Python_Design-Patterns/facade/point.py
import math
from facade.point_carte import PointCarte
from facade.point_polar import PointPolar


class Point:

    def __init__(self, x, y):
        pc = PointCarte(x, y)

    def __str__(self):
        return self.pc.__str__()

    def rotate(self, angle, Point):
        x = self.pc.x() - Point.pc.x()
        y = self.pc.y() - Point.pc.y()

        pp = PointPolar(math.sqrt(math.pow(x, 2)) + math.pow(y, 2),
                        math.atan2(y, x) * 180 / math.pi)

        pp.rotate(angle)
        print(f'PointPolar is {pp}')
        my_string = pp.__str__()
        i = my_string.index('@')
        r = float(my_string.string[1:i])
        a = float(my_string.string[i + 1:my_string.__len__() - 1])

        pc = PointCarte(r * math.cos(a * math.pi / 180) + Point.pc.x(),
                        r * math.sin(a * math.pi / 180) + Point.pc.y())
